blosc/blosc.c

```diff
@@ -2117,6 +2117,20 @@ void blosc_set_splitmode(int mode)
   g_splitmode = mode;
 }
 
+/* Child process global context is an invalid state and threads are no longer
+ * valid.  Re-init a new context. */
+void blosc_fork_child() {
+  if (!g_initlib) return;
+
+  /* trylock, fork may have left global mutex locked in another thread */
+  pthread_mutex_trylock(&global_comp_mutex);
+
+  g_global_context = (struct blosc_context*)my_malloc(sizeof(struct blosc_context));
+  g_global_context->threads_started = 0;
+
+  pthread_mutex_unlock(&global_comp_mutex);
```

Comment:
You cannot unlock a mutex that you don't own. I think you have to reinitialize the lock instead.

Reply:
Thank you for the review. This uncovered an error causing in duplicate atfork handler registration, which I've resolved.

Reply:
I believe the most-portable solution here is to convert `global_comp_mutex` to a `pthread_mutex_t*`, then fully de-initialize the global context in the child post-fork. Subsequent calls to the `blosc_*` interfaces will trigger re-init of the global context.

Reply:
Very interesting read, the history of phtread_atfork! Not only does POSIX forbid calling async-signal-unsafe functions in post-fork handler, but also during the whole lifetime of the child between fork and exec. Which means your test case is invalid code since blosc API isn't async-singnal-safe.

Reply:
Yes, it makes sense. Let's go on with this patch.